api/src/main/kotlin/com/google/devtools/ksp/processing/CodeGenerator.kt

```diff
@@ -32,22 +32,38 @@ interface CodeGenerator {
     /**
      * Creates a file which is managed by [CodeGenerator]
      *
+     * Sources of corresponding [KSNode]s which are obtained directly from [Resolver] need to be specified.
+     * Namely, the containing files of those [KSNode]s who are obtained from:
+     *   * [Resolver.getAllFiles]
+     *   * [Resolver.getSymbolsWithAnnotation]
+     *   * [Resolver.getClassDeclarationByName]
+     *
+     * Instead of requiring processors to specify all source files which are relevant in generating the given output,
+     * KSP traces dependencies automatically and only need to know those sources that only processors know what they
+     * are for. If a [KSFile] is indirectly obtained through other [KSNode]s, it hasn't to be specified for the given
+     * output, even if its contents contribute to the generation of the output.
+     *
+     * For example, a processor generates an output `O` after reading class `A` in `A.kt` and class `B` in `B.kt`,
+     * where `A` extends `B`. The processor got `A` by [Resolver.getSymbolsWithAnnotation] and then got `B` by
+     * [KSClassDeclaration.superTypes] from `A`. Because the inclusion of `B` is due to `A`, `B.kt` needn't to be
+     * specified in [dependencies] for `O`. Note that specifying `B.kt` in this case doesn't hurt, it is only unnecessary.
+     *
+     * @param dependencies are [KSFile]s from which this output is built. Only those that are obtained directly
+     *                     from [Resolver] are required.
```

Comment:
curious, if the generated code references the file, does it still need to be provided?

e.g. if I query `getSymboldsAnnotatedWith`, get interface `Foo` and generate `FooImpl : Foo`, do I still need to specify the source file for `Foo` or does it automatically get captured?

Reply:
Yes, it still is needed. Current incremental + multiple-round plan is to leave generated code alone when computing dependencies; I.e., not rely on them at all.